### ISSUE TITLE
Not able anymore to use Magento by specifying base_url placeholder

### ISSUE THREAD

**Issue by romainruaud** [magento/magento2#10693]
We are running Magento automatical deployments on a **containers architecture**.

URL of the Magento instances are generated "on the fly" (like magento-test-23qsqd456.domain.com) and therefore we process an automatical installation on these containers via something like this (configuration is coming from environment variables) : 

```
    php bin/magento setup:install                  \
        --admin-firstname=$ADMIN_FIRSTNAME         \
        --admin-lastname=$ADMIN_LASTNAME           \
        --admin-email=$ADMIN_EMAIL                 \
        --admin-user=$ADMIN_USER                   \
        --admin-password=$ADMIN_PASSWORD           \
        --backend-frontname=$BACKEND_FRONTNAME     \
        --db-host=$MYSQL_HOSTNAME                  \
        --db-name=$MYSQL_DATABASE                  \
        --db-user=$MYSQL_USER                      \
        --db-password=$MYSQL_PASSWORD              \
        --base-url={{base_url}}                    \
        --use-rewrites=1                           \
        --cleanup-database                         \
        --magento-init-params=MAGE_MODE=production

```

By using these parameters, we were previously able to have a running Magento instance on a dynamically created URL (eg : magento-23qs2d3qsd.test.domain.com)

This was due to the fact that Magento is retrieving the base url via ```getDistroBaseUrl``` as used here : https://github.com/magento/magento2/blob/develop/app/code/Magento/Store/Model/Store.php#L648

But since this commit (https://github.com/magento/magento2/commit/4bfd90d324ca6b743b04db5963bcb3aea44617e8), the value is always explicitely replaced by "http://localhost/" (see here : https://github.com/magento/magento2/blob/develop/app/code/Magento/Store/Model/Config/Placeholder.php#L94) which is irrelevant and leads to Magento redirecting to "http://localhost/" when accessing it.

I note that there is a comment saying "localhost is replaced for cli requests, for http requests method getDistroBaseUrl is used" which is obviously **wrong** because it **does** affect HTTP requests.

### Preconditions

1. Magento 2.1.8 or Magento 2.2
2. PHP 7.0 (but should be the same on 5.6)

### Steps to reproduce

1. Install Magento on a container or a machine having its own domain name (eg test-magento.mytestingserver.com )
2. Use {{base_url}} placeholder when installing
3. Access test-magento.mytestingserver.com

### Expected result

1. Magento should render the page by dynamically using the request URL as base_url

### Actual result

1. [Screenshot, logs]

Redirection to http://localhost/?SID=22u1m71ajo1fup8aq31m4kj641

**Comment by misha-kotov** [comment]
Thanks for reporting, internal issue is MAGETWO-71893

**Comment by magento-engcom-team** [comment]
Hi @romainruaud 

This issue seems to be already fixed in 2.2. 

You may use commit be34d52943b1ad60f90df9c2922e362997dbd536 as a reference.

Thank you
